AX: When computing AXCoreObject::unignoredChildren, we re-retrieve the same parent object numerous times (sometimes thousands of times), hurting performance

twilco opened the pull request:
#### c976b42339408e0d64e484208f62fac9e9489e5e
<pre>
AX: When computing AXCoreObject::unignoredChildren, we re-retrieve the same parent object numerous times (sometimes thousands of times), hurting performance
<a href="https://bugs.webkit.org/show_bug.cgi?id=298251">https://bugs.webkit.org/show_bug.cgi?id=298251</a>
<a href="https://rdar.apple.com/159684099">rdar://159684099</a>

Reviewed by Joshua Hoffman.

One &quot;phase&quot; of computing AXCoreObject::unignoredChildren is iterating through the siblings of an object to find one that
is eligible. It&apos;s a bit more complicated than that, as nextInPreOrder may cause us to go down a level in the tree, but
often we iterate through siblings at the same level hundreds or thousands of times. Prior to this commit, this sibling
iteration happened by calling AXCoreObject::nextSiblingIncludedIgnored(), which re-retrieved the parent object every time.
This has costs — chasing a pointer to the object&apos;s AXIsolatedTree, member variable read to get the parent ID, a hashmap
lookup in the tree, and a +1 and -1 ref-count churn. With a function as hot as unignoredChildren, this is a gigantic cost.

This commit &quot;inlines&quot; the relevant bits of AXCoreObject::nextSiblingIncludedIgnored into AXCoreObject::unignoredChildren,
allowing us to implement a mechanism that caches a descendant&apos;s parent and siblings.

I took a sample before and after this change with 90 seconds of VoiceOver navigation on html.spec.whatwg.org:

Before: 10210 samples spent in AXCoreObject::unignoredChildren
After: 4641 samples spent in AXCoreObject::unignoredChildren

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::unignoredChildren):
* Source/WebCore/accessibility/AXCoreObject.h:

Canonical link: <a href="https://commits.webkit.org/299474@main">https://commits.webkit.org/299474@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5280936d263dab899ecc00ac9ebeeeb923232672

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119080 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29412 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125292 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71143 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39457 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47343 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90415 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122033 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31453 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106746 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70876 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30515 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24866 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68947 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25047 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128321 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34750 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99030 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98810 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25129 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44270 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22268 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42566 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45857 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51536 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45324 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48672 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47009 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/298ba795b4fcbc08c13d4f7ab339b71b3e4416bc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50189

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5280936d263dab899ecc00ac9ebeeeb923232672. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50189

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -284,25 +284,45 @@ AXCoreObject::AccessibilityChildrenVector AXCoreObject::unignoredChildren(bool u
+                siblings = &parent->childrenIncludingIgnored();
+                currentSiblingIndex = descendant->indexInSiblings(*siblings);
+            }
+            ++(*currentSiblingIndex);
```
Fixed in latest commit!

twilco committed:
AX: When computing AXCoreObject::unignoredChildren, we re-retrieve the same parent object numerous times (sometimes thousands of times), hurting performance
https://bugs.webkit.org/show_bug.cgi?id=298251
rdar://159684099

Reviewed by Joshua Hoffman.

One "phase" of computing AXCoreObject::unignoredChildren is iterating through the siblings of an object to find one that
is eligible. It's a bit more complicated than that, as nextInPreOrder may cause us to go down a level in the tree, but
often we iterate through siblings at the same level hundreds or thousands of times. Prior to this commit, this sibling
iteration happened by calling AXCoreObject::nextSiblingIncludedIgnored(), which re-retrieved the parent object every time.
This has costs — chasing a pointer to the object's AXIsolatedTree, member variable read to get the parent ID, a hashmap
lookup in the tree, and a +1 and -1 ref-count churn. With a function as hot as unignoredChildren, this is a gigantic cost.

This commit "inlines" the relevant bits of AXCoreObject::nextSiblingIncludedIgnored into AXCoreObject::unignoredChildren,
allowing us to implement a mechanism that caches a descendant's parent and siblings.

I took a sample before and after this change with 90 seconds of VoiceOver navigation on html.spec.whatwg.org:

Before: 10210 samples spent in AXCoreObject::unignoredChildren
After: 4641 samples spent in AXCoreObject::unignoredChildren

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::unignoredChildren):
* Source/WebCore/accessibility/AXCoreObject.h:

Canonical link: https://commits.webkit.org/299474@main

webkit-commit-queue commented:
Committed 299474@main (c976b4233940): <https://commits.webkit.org/299474@main>

Reviewed commits have been landed. Closing PR #50189 and removing active labels.
